modules/tlsinfo.cpp

```diff
@@ -0,0 +1,329 @@
+ in a user friendly dialog. */
+/* As-is customary, the chain starts with the peer certificate and
+ and ends with the root certificate authority certificate. */
+/* The certificate chain is sent using the BATCH command. */
+/* Each certificate is surrounded in its own nested BATCH within
+ a single global BATCH for the whole session. */
+
+/* Example response:
+
+	>> :znc.in BATCH +1f5fdd znc.in/tlsinfo
+	>> @batch=1f5fdd :znc.in BATCH +6f7a1a znc.in/tlsinfo-certificate
+	>> @batch=6f7a1a :znc.in CERTINFO ExampleUser :-----BEGIN CERTIFICATE-----
+	>> @batch=6f7a1a :znc.in CERTINFO ExampleUser :...
+	>> @batch=6f7a1a :znc.in CERTINFO ExampleUser :-----END CERTIFICATE-----
+	>> @batch=1f5fdd :znc.in BATCH -6f7a1a
+	>> @batch=1f5fdd :znc.in BATCH +31fbfc znc.in/tlsinfo-certificate
+	>> @batch=31fbfc :znc.in CERTINFO ExampleUser :-----BEGIN CERTIFICATE-----
+	>> @batch=31fbfc :znc.in CERTINFO ExampleUser :...
+	>> @batch=31fbfc :znc.in CERTINFO ExampleUser :-----END CERTIFICATE-----
+	>> @batch=1f5fdd :znc.in BATCH -31fbfc
+	>> :znc.in BATCH -1f5fdd
+ */
+
+#include <znc/Modules.h>
+#include <znc/IRCNetwork.h>
+#include <znc/IRCSock.h>
+
+static const char *TlsInfoCap = "znc.in/tlsinfo";
+
+static const char *TlsInfoBatchGlobalType = "znc.in/tlsinfo";
+static const char *TlsInfoBatchChildType = "znc.in/tlsinfo-certificate";
+
+class CTlsInfoMod : public CModule
+{
+public:
+
+#ifndef HAVE_LIBSSL
+	MODCONSTRUCTOR(CTlsInfoMod) {}
+
+	bool OnLoad(const CString&, CString& sMessage) {
+		sMessage = "Module built against install of ZNC that lacks SSL support.";
+
+		return false;
+	}
+
+#else // HAVE_LIBSSL
+
+	MODCONSTRUCTOR(CTlsInfoMod) {
+		AddHelpCommand();
+		AddCommand("Cert", static_cast<CModCommand::ModCmdFunc>(&CTlsInfoMod::PrintCertificateCommand), "[details]", "View certificate information for the active connection. Append 'details' to the 'cert' command ('cert details') to include the entire certificate chain in output.");
+		AddCommand("Cipher", static_cast<CModCommand::ModCmdFunc>(&CTlsInfoMod::PrintCipherCommand), "", "View the protocol and cipher suite used for the active connection.");
+		AddCommand("Send-Data", static_cast<CModCommand::ModCmdFunc>(&CTlsInfoMod::SendCertificateCommand), "", "Send certificate information for the active connection in an easily parsable format for application developers.");
+	}
+
+	void OnClientCapLs(CClient *mClient, SCString &mCaps) override
+	{
+		mCaps.insert(TlsInfoCap);
+	}
+
+	bool IsClientCapSupported(CClient *mClient, const CString &mCap, bool mState) override
+	{
+		return mCap.Equals(TlsInfoCap);
+	}
+
+	void PrintCipherCommand(const CString &mLine)
+	{
+		CClient *mClient;
+
+		SSL *mSSLObject;
+
+		if (GetRelevantObjects(nullptr, &mClient, nullptr, &mSSLObject) == false) {
+			return;
+		}
+
+		/* Determine the protocol used for the connection */
+		int mSSLVersion = SSL_version(mSSLObject);
+
+		CString mSSLVersionString;
+
+		if (mSSLVersion == TLS1_2_VERSION) {
```

Comment:
Old enough openssl doesn't define too new symbols. I don't know though what minimum openssl version should be required.